---
title: "Gene set enrichment analysis -- MAGMA"
author: "XSun"
date: "2024-05-29"
output:
  workflowr::wflow_html:
    code_folding: hide
    toc: true
---

# Introduction

## Gene set overview

We used MAGMA to do the gene set enrichment analysis for 15 gene sets below. Since MAGMA works on ENTREZ ID, we converted the gene symbols into ENTREZID using `EnsDb.Hsapiens.v75`

```{r  message=FALSE}
library(ggplot2)
library(tidyr)
library(reshape2)


load("/project2/xinhe/xsun/neuron_stimulation/5.magma/data/all_set_annotated.rdata")
DT::datatable(all_set,caption = htmltools::tags$caption( style = 'caption-side: left; text-align: left; color:black;  font-size:150% ;','Annotated gene sets'),options = list(pageLength = 5) )

```


```{r message=FALSE}
# Count the number of genes in each cluster

load("/project2/xinhe/xsun/neuron_stimulation/5.magma/data/geneset_gene_count.rdata")

DT::datatable(gene_counts,caption = htmltools::tags$caption( style = 'caption-side: left; text-align: left; color:black;  font-size:150% ;','The number of genes in each cluster'),options = list(pageLength = 5) )

gene_counts_long <- gene_counts %>%
  pivot_longer(cols = c(count_GENESYMBOL, count_ENTREZID), names_to = "Type", values_to = "Count")

# Plot the barplot
ggplot(gene_counts_long, aes(x = as.factor(Cluster), y = Count, fill = Type)) +
  geom_bar(stat = "identity", position = position_dodge()) +
  labs(x = "Cluster", y = "Count", fill = "Type", title = "Gene Counts by Cluster") +
  theme_minimal()
```

## MAGMA settings

We ran MAGMA following https://cloufield.github.io/GWASTutorial/09_Gene_based_analysis/#download-reference-files 

- Annotate SNPs

To capture the variants in the regulatory regions, we set a 50 kb upstream and 50kb downstream window for each gene.

- Gene-based analysis

GWAS are required in this step. The traits included are: 

```{r message=FALSE}
# Count the number of genes in each cluster

gwas <- read.table("/project2/xinhe/xsun/neuron_stimulation/5.magma/data/gwas_source_update.txt",header = T)

DT::datatable(gwas,caption = htmltools::tags$caption( style = 'caption-side: left; text-align: left; color:black;  font-size:150% ;','GWAS included'),options = list(pageLength = 5) )

```

**MD stands for Major depression, BP is bipolar, PK is Parkinson, NS is Neuroticism score**

- Gene-set level analysis

The gene-set file is `/project2/xinhe/xsun/neuron_stimulation/5.magma/data/genesets_entrezid.txt`.

# Results

```{r message=FALSE}
# Count the number of genes in each cluster

load("/project2/xinhe/xsun/neuron_stimulation/5.magma/results/geneset_p.rdata")

DT::datatable(p_set,caption = htmltools::tags$caption( style = 'caption-side: left; text-align: left; color:black;  font-size:150% ;','Gene-set analysis p-values from MAGMA'),options = list(pageLength = 5) )

pvalues_long <- melt(p_set, varnames = c("Set", "Condition"), value.name = "PValue")

# Plot the heatmap with conditional coloring
ggplot(pvalues_long, aes(x = Condition, y = Set, fill = PValue < 0.05)) +
  geom_tile(color = "white") +
  scale_fill_manual(values = c("TRUE" = "red", "FALSE" = "blue"), labels = c("P-value > 0.05", "P-value < 0.05")) +
  labs(title = "Heatmap of P-values", x = "Traits", y = "Gene Sets", fill = "p from MAGMA geneset analysis") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

```


Then do a multiple testing correction for all the tests (15 gene sets x 7 traits) using q-values. 

The q-value is computed by `qvalue(p_values)` from `qvalue` package

```{r message=FALSE}
# Count the number of genes in each cluster

load("/project2/xinhe/xsun/neuron_stimulation/5.magma/results/q_values_matrix.rdata")

DT::datatable(q_values_matrix,caption = htmltools::tags$caption( style = 'caption-side: left; text-align: left; color:black;  font-size:150% ;','Gene-set analysis p-values from MAGMA'),options = list(pageLength = 5) )

qvals_long <- melt(q_values_matrix, varnames = c("Set", "Condition"), value.name = "qval")

# Plot the heatmap with conditional coloring
ggplot(qvals_long, aes(x = Condition, y = Set, fill = qval < 0.3)) +
  geom_tile(color = "white") +
  scale_fill_manual(values = c("TRUE" = "red", "FALSE" = "blue"), labels = c("q-values > 0.3", "q-values < 0.3")) +
  labs(title = "Heatmap of q-values", x = "Traits", y = "Gene Sets", fill = "q-values from MAGMA geneset analysis") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

```

<!-- ```{r message=FALSE} -->
<!-- # Count the number of genes in each cluster -->

<!-- load("/project2/xinhe/xsun/neuron_stimulation/5.magma/results/fdr_matrix.rdata") -->

<!-- DT::datatable(fdr_matrix,caption = htmltools::tags$caption( style = 'caption-side: left; text-align: left; color:black;  font-size:150% ;','Gene-set analysis p-values from MAGMA'),options = list(pageLength = 5) ) -->

<!-- # Plot the heatmap with conditional coloring -->
<!-- FDRs_long <- melt(fdr_matrix, varnames = c("Set", "Condition"), value.name = "FDR") -->

<!-- # Plot the heatmap with conditional coloring -->
<!-- ggplot(FDRs_long, aes(x = Condition, y = Set, fill = FDR < 0.3)) + -->
<!--   geom_tile(color = "white") + -->
<!--   scale_fill_manual(values = c("TRUE" = "red", "FALSE" = "blue"), labels = c("FDR > 0.3", "FDR < 0.3")) + -->
<!--   labs(title = "Heatmap of FDR", x = "Traits", y = "Gene Sets", fill = "FDR from MAGMA geneset analysis") + -->
<!--   theme_minimal() + -->
<!--   theme(axis.text.x = element_text(angle = 45, hjust = 1)) -->

<!-- ``` -->

